### PR TITLE
Serializable `Stream` object

### PR DESCRIPTION
**Pull request by tylerganter**
Refactoring as serializable stream class interface.

`MjpegStream` and `ImageStream` have also now been implemented ~~however the `get_live_stream_url` is not implemented for either. I am going to need to coordinate with @benjaminpkane or @lethosor for how these streams will be exposed via the API.~~ This is done now as well

Also, for @ehofesmann , I wasn't really sure what things were heavily custom to a given stream source and what were general, so I didn't think too hard about it. The structure is there and it should now be easy to make new subclasses and move common functionality to parent classes, etc. All that fun stuff of OOP

### CODE REVIEW

File: config/streams/dublin.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "type": "pandemic51.core.streaming.M3U8Stream",
+    "webpage": "https://www.earthcam.com/world/ireland/dublin/?cam=templebar",
+    "chunk_path": "https://d3o4twxzdiwvsf.cloudfront.net/fecnetwork/4054.flv/chunklist.m3u8",
+    "GMT": "0"
+}
```
Comment (ehofesmann) [inline config/streams/dublin.json:6]: Is it ok for some of these streams to be missing `stream_name`

File: pandemic51/core/streaming.py [hunk 2]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
+        urls = get_img_urls(self.webpage)
+        filtered_urls = [u for u in urls if self.url_filter in u]
+
+        if not filtered_urls:
+            raise Exception("No URLs found for webpage: %s" % self.webpage)
+
+        # we only need one!
+        url = filtered_urls[0]
+
+        # Get the large version of the image instead of the thumbnail
```
Comment (ehofesmann) [inline pandemic51/core/streaming.py:429]: This is really only applicable to the Ypsilanti livestream, the rest of this class seems pretty general though

**Review by ehofesmann — APPROVED** [review]
Just a couple of questions, looks pretty good though

File: pandemic51/core/streaming.py [hunk 3]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
 
 
-def download_and_store(stream_name, outdir, width=None, height=None):
-    '''Downloads an image from the latest stream, and add it to the database.
+class Stream(etas.Serializable):
+    '''Abstract class for streams of various types providing functionality for
+    downloading clips and images from the stream, getting the latest stream URL,
+    etc.
+    '''
+    def __init__(self, stream_name, GMT):
```
Comment (brimoor) [inline pandemic51/core/streaming.py:119]: nit: style guide says no capital letters in variable names (GMT)

File: pandemic51/core/streaming.py [hunk 4]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
+        Args:
+            stream_name: name of the stream
+
+        Returns:
+            path the the Stream object serialized on disk
+        '''
+        return os.path.join(panc.STREAMS_DIR, stream_name + ".json")
+
+    @staticmethod
+    def get_stream_names():
```
Comment (brimoor) [inline pandemic51/core/streaming.py:143]: convert to a public module method? It's about all streams, not one particular stream

File: pandemic51/core/streaming.py [hunk 5]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
+        Returns:
+            a Stream instance
+        '''
+        d = etas.read_json(path)
 
-    Returns:
-        image_path: path the the downloaded image on disk
-        dt: datetime object of when the image was downloaded
+        if "stream_name" not in d:
+            d["stream_name"] = os.path.splitext(os.path.basename(path))[0]
```
Comment (brimoor) [inline pandemic51/core/streaming.py:224]: Why allow `stream_name` to be absent? Seems like more confusion than utility. @ehofesmann asked about this, for example

File: pandemic51/core/streaming.py [hunk 6]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
+        '''
+        return cls.from_json(cls.stream_path(stream_name))
+
+    @classmethod
+    def _from_dict(cls, d):
+        raise NotImplementedError("Subclass must implement")
+
+
+class M3U8Stream(Stream):
+    '''A Stream class that reads URIs from an M3U8 chunk path'''
```
Comment (brimoor) [inline pandemic51/core/streaming.py:246]: nit: space after class docstring

File: config/streams/dublin.json (hunk 1 above)
Comment (tylerganter) [reply config/streams/dublin.json:6]: yes, it's automatically populated if it's missing

File: pandemic51/core/streaming.py (hunk 2 above)
Comment (tylerganter) [reply pandemic51/core/streaming.py:429]: I'm going to leave it until we add another `ImageStream` if you're okay with that

File: pandemic51/core/streaming.py [hunk 7]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
+
+        try:
+            urllib.request.urlopen(url)
+        except urllib.error.HTTPError:
+            self.update_stream_chunk_path()
+            url = self.chunk_path
+
+        if "videos2archives" in url:
+            url = (
+                    "https://pdi-service.voxel51.com/stream-archive/" +
```
Comment (brimoor) [inline pandemic51/core/streaming.py:263]: Consider refactoring these magic URLs to config constants? Others have been done like that

File: pandemic51/core/streaming.py [hunk 8]
```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
...
+                    url.split(".com/")[1]
+            )
+        elif "earthcam" in url:
+            url = (
+                    "https://pdi-service.voxel51.com/stream/" +
```
Comment (brimoor) [inline pandemic51/core/streaming.py:268]: nit: 4 space indent

**Review by brimoor — COMMENTED** [review]
Looks like a nice refactor! Left a few small comments

**Pull request merged**